src/Grid.hpp

```diff
@@ -0,0 +1,90 @@
+//
+//  Grid.hpp
+//  Game of Life
+//
+//  Created by Alex Frasca on 01/05/2023.
+//
+#pragma once
+#ifndef Grid_hpp
+#define Grid_hpp
+
+#include <SFML/Graphics.hpp>
+#include <vector>
+#include <cmath>
+#include <functional>
+
+namespace conway {
+
+//several values so we can draw nice colors as the cells die instead of just on/off
+enum CELL_STATE {
+    CELL_OFF = 0,
+    CELL_LOW = 1,
+    CELL_MEDIUM = 2,
+    CELL_HIGH = 3,
+    CELL_ON = 4
+};
+const float DEFAULT_SPACING = 16.0;
+
+struct Cell {
+    
+    Cell() {};
+    Cell(sf::Vector2<int> idx, sf::Vector2<float> pos, float spc, CELL_STATE st);
+    
+    void tick();
+    void setState();
+    // for a dead cell, use a frame-wise decrement for animation/visualization purposes
+    void decrement();
+    // only for when the size or spacing of the grid is changed, not for each tick
+    void updatePosition();
+    
+    // Cell member vars
+    sf::Vector2<int> index;
+    sf::Vector2<float> position;
+    // for drawing the cell
+    sf::ConvexShape drawable;
+    
+    float spacing;
+    
+    // bools for determining the next state of the grid. we need this otherwise grid cells
+    // will update as we tick them resulting in undesired neighbor calculations for future cells
+    bool kill_next_cycle;
+    bool activate_next_cycle;
+    
+    CELL_STATE state;
+
+};
+
+class Grid {
+public:
+    
+    Grid();
+    Grid(uint32_t width, uint32_t height);
+    
+    uint32_t grid_width;
+    uint32_t grid_height;
+    
+    // how far apart are the grid elements?
+    float spacing;
+    
+    //the cells of the grid
+    std::vector<Cell> cells;
```

Comment:
Should be private to the class, otherwise it can't guarantee maintenance of associated invariants

Reply:
just the vector? or should other vars be private as well?

Reply:
Haven't gone through the logic to be able to answer that accurately, but in general anything that if poked at from outside the class could mess up the state of the object, is an invariant the class needs to maintain, and thus cannot be public.